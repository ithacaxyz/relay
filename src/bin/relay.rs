//! # Ithaca Relay
//!
//! A relay service that sponsors transactions for EIP-7702 accounts.

use clap::Parser;
use relay::{
    cli::*,
    otlp::{OtelConfig, OtelGuard},
};
use tracing::debug;
use tracing_subscriber::prelude::*;

fn init_tracing_subscriber() -> Option<OtelGuard> {
    let registry = tracing_subscriber::registry().with(
        tracing_subscriber::fmt::layer()
            .with_filter(tracing_subscriber::filter::EnvFilter::from_default_env()),
    );

    if let Some(cfg) = OtelConfig::load() {
        let guard = cfg.provider();
        registry.with(guard.layer()).init();
        Some(guard)
    } else {
        registry.init();
        None
    }
}

#[tokio::main]
async fn main() {
    // Enable backtraces unless a RUST_BACKTRACE value has already been explicitly provided.
    if std::env::var_os("RUST_BACKTRACE").is_none() {
        unsafe { std::env::set_var("RUST_BACKTRACE", "1") };
    }

    let _guard = init_tracing_subscriber();
    if _guard.is_some() {
        debug!("opentelemetry initialized");
    }

    let args = Args::parse();
    if let Err(err) = args.run().await {
        eprintln!("Error: {err:?}");
        std::process::exit(1);
    }
}
