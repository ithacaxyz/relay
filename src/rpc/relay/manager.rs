use crate::{
    asset::AssetInfoServiceHandle,
    chains::Chains,
    config::QuoteConfig,
    error::{AuthError, RelayError},
    price::PriceOracle,
    storage::StorageApi,
    types::{Account, FeeTokens, VersionedContracts},
};
use alloy::{
    eips::eip7702::constants::EIP7702_DELEGATION_DESIGNATOR,
    primitives::{Address, Bytes, ChainId},
    providers::{DynProvider, Provider},
    rpc::types::state::StateOverridesBuilder,
};
use tracing::instrument;

#[async_trait::async_trait]
pub trait Manager {
    /// Get chains
    fn chains(&self) -> &Chains;

    /// Get contracts
    fn contracts(&self) -> &VersionedContracts;

    /// Get orchestrator address
    fn orchestrator(&self) -> Address {
        self.contracts().orchestrator.address
    }

    /// Previously deployed orchestrators.
    fn legacy_orchestrators(&self) -> impl Iterator<Item = Address> {
        self.contracts().legacy_orchestrators.iter().map(|c| c.address)
    }

    /// Get delegation implementation address
    fn delegation_implementation(&self) -> Address {
        self.contracts().delegation_implementation.address
    }

    /// Previously deployed delegation implementations.
    fn legacy_delegations(&self) -> impl Iterator<Item = Address> {
        self.contracts().legacy_delegations.iter().map(|c| c.address)
    }

    /// The simulator address.
    fn simulator(&self) -> Address {
        self.contracts().simulator.address
    }

    /// Get fee recipient
    fn fee_recipient(&self) -> Address;

    /// Get priority fee percentile
    fn priority_fee_percentile(&self) -> f64;

    /// Get quote config
    fn quote_config(&self) -> &QuoteConfig;

    /// Get fee tokens
    fn fee_tokens(&self) -> &FeeTokens;

    /// Get price oracle
    fn price_oracle(&self) -> &PriceOracle;

    /// Get asset info
    fn asset_info(&self) -> &AssetInfoServiceHandle;

    /// Get storage
    fn storage(&self) -> &dyn StorageApi;

    /// Get a provider for the given chain ID
    fn provider(&self, chain_id: ChainId) -> Result<DynProvider, RelayError> {
        self.chains()
            .get(chain_id)
            .map(|chain| chain.provider.clone())
            .ok_or(RelayError::UnsupportedChain(chain_id))
    }

    /// Checks if the orchestrator is supported.
    fn is_supported_orchestrator(&self, orchestrator: &Address) -> bool {
        self.orchestrator() == *orchestrator
            || self.legacy_orchestrators().any(|c| c == *orchestrator)
    }

    /// Checks if the account has a supported delegation implementation. If so, returns it.
    async fn has_supported_delegation<P: Provider + Clone>(
        &self,
        account: &Account<P>,
    ) -> Result<Address, RelayError> {
        let address = self.get_delegation_implementation(account).await?;
        if self.delegation_implementation() == address
            || self.legacy_delegations().any(|c| c == address)
        {
            return Ok(address);
        }
        Err(AuthError::InvalidDelegation(address).into())
    }

    /// Ensures the account has the latest delegation implementation. Otherwise, returns error.
    async fn ensure_latest_delegation<P: Provider + Clone>(
        &self,
        account: &Account<P>,
    ) -> Result<(), RelayError> {
        let address = self.has_supported_delegation(account).await?;
        if self.delegation_implementation() != address {
            return Err(AuthError::InvalidDelegation(address).into());
        }
        Ok(())
    }

    /// Returns the delegation implementation address from the requested account.
    ///
    /// It will return error if the delegation proxy is invalid.
    #[instrument(skip_all)]
    async fn get_delegation_implementation<P: Provider + Clone>(
        &self,
        account: &Account<P>,
    ) -> Result<Address, RelayError> {
        if let Some(delegation) = account.delegation_implementation().await? {
            return Ok(delegation);
        }

        // Attempt to retrieve the delegation proxy from storage, since it might not be
        // deployed yet.
        let Some(stored) = self.storage().read_account(&account.address()).await? else {
            return Err(RelayError::Auth(AuthError::EoaNotDelegated(account.address()).boxed()));
        };

        let address = account.address();
        let account = account.clone().with_overrides(
            StateOverridesBuilder::default()
                .with_code(
                    address,
                    Bytes::from(
                        [
                            &EIP7702_DELEGATION_DESIGNATOR,
                            stored.signed_authorization.address().as_slice(),
                        ]
                        .concat(),
                    ),
                )
                .build(),
        );

        account.delegation_implementation().await?.ok_or_else(|| {
            RelayError::Auth(
                AuthError::InvalidDelegationProxy(*stored.signed_authorization.address()).boxed(),
            )
        })
    }
}
