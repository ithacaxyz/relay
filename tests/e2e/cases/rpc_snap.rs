use crate::e2e::{
    AuthKind, ExpectedOutcome, MockErc20, TxContext, await_calls_status,
    cases::{upgrade_account_eagerly, upgrade_account_lazily},
    environment::{Environment, EnvironmentConfig},
};
use alloy::{
    primitives::{Address, B256, Bytes, U64, U256, b256},
    providers::Provider,
    sol_types::SolCall,
    uint,
};
use relay::{
    rpc::RelayApiClient,
    signers::Eip712PayLoadSigner,
    types::{
        Call, CallPermission, IERC20, KeyType, KeyWith712Signer,
        rpc::{
            AddFaucetFundsParameters, BundleId, GetAssetsParameters, GetAuthorizationParameters,
            GetKeysParameters, Meta, Permission, PrepareCallsCapabilities, PrepareCallsParameters,
            PrepareUpgradeAccountParameters, RequiredAsset, SendPreparedCallsParameters,
            UpgradeAccountCapabilities, UpgradeAccountParameters, UpgradeAccountSignatures,
            VerifySignatureParameters,
        },
    },
};
use std::str::FromStr;

const ADMIN_KEY: B256 = b256!("0x013c83ce5b08455e0505392b45f6d5effc7671e6f69108268ee4fedae1df72c8");
const SESSION_KEY: B256 =
    b256!("0xe1d7a91c93712db419ab647152bc1ae8739aca2f1bc8a6220d47fcf4a541a73c");

#[tokio::test]
async fn test_health() -> eyre::Result<()> {
    let env = Environment::setup().await?;

    let response = env.relay_endpoint.health().await?;
    insta::assert_json_snapshot!(response, {
        ".status" => reduction_from_str::<String>("status"),
        ".version" => reduction_from_str::<String>("version"),
    });

    Ok(())
}

#[tokio::test]
async fn test_get_capabilities() -> eyre::Result<()> {
    let env = Environment::setup().await?;

    let mut response =
        env.relay_endpoint.get_capabilities(Some(vec![U64::from(env.chain_id())])).await?;

    for (_, caps) in response.0.iter_mut() {
        caps.contracts.legacy_orchestrators.sort();
        caps.contracts.legacy_delegations.sort();
        caps.fees.tokens.sort_by_key(|token| token.uid.clone());
    }

    insta::assert_json_snapshot!(response);

    Ok(())
}

#[tokio::test]
async fn test_get_keys() -> eyre::Result<()> {
    let env = Environment::setup().await?;

    let admin_key = KeyWith712Signer::mock_admin_with_key(KeyType::Secp256k1, ADMIN_KEY)?.unwrap();
    let session_key = KeyWith712Signer::mock_session_with_key(KeyType::P256, SESSION_KEY)?
        .unwrap()
        .with_permissions(vec![Permission::Call(CallPermission {
            to: env.erc20,
            selector: MockErc20::transferCall::SELECTOR.into(),
        })]);

    upgrade_account_eagerly(&env, &[admin_key.to_authorized()], &admin_key, AuthKind::Auth).await?;

    // Add session key
    TxContext {
        authorization_keys: vec![&session_key],
        expected: ExpectedOutcome::Pass,
        key: Some(&admin_key),
        ..Default::default()
    }
    .process(0, &env)
    .await?;

    let response = env
        .relay_endpoint
        .get_keys(GetKeysParameters { address: env.eoa.address(), chain_ids: vec![] })
        .await?;
    insta::assert_json_snapshot!(response);

    Ok(())
}

#[tokio::test]
async fn test_get_assets() -> eyre::Result<()> {
    let env = Environment::setup().await?;

    let mut response = env
        .relay_endpoint
        .get_assets(GetAssetsParameters { account: env.eoa.address(), ..Default::default() })
        .await?;

    for (_, assets) in response.0.iter_mut() {
        assets.sort_by_key(|asset| asset.address);
    }

    insta::assert_json_snapshot!(response);

    Ok(())
}

#[tokio::test]
async fn test_prepare_calls() -> eyre::Result<()> {
    let config =
        EnvironmentConfig { num_chains: 2, fee_recipient: Address::ZERO, ..Default::default() };
    let env = Environment::setup_with_config(config.clone()).await?;

    // Create a key for signing
    let key = KeyWith712Signer::mock_admin_with_key(KeyType::Secp256k1, ADMIN_KEY)?.unwrap();

    // Account upgrade deployed onchain.
    upgrade_account_lazily(&env, &[key.to_authorized()], AuthKind::Auth).await?;

    let balance =
        IERC20::new(env.erc20, env.provider_for(1)).balanceOf(env.eoa.address()).call().await?
            / uint!(2_U256);
    let response = env
        .relay_endpoint
        .prepare_calls(PrepareCallsParameters {
            calls: vec![Call::transfer(env.erc20, Address::ZERO, uint!(1_U256))],
            chain_id: env.chain_id_for(0),
            from: Some(env.eoa.address()),
            capabilities: PrepareCallsCapabilities {
                authorize_keys: Default::default(),
                meta: Meta { fee_token: env.erc20, fee_payer: None, nonce: Some(U256::ZERO) },
                pre_calls: Default::default(),
                pre_call: Default::default(),
                required_funds: vec![RequiredAsset::new(env.erc20, balance)],
                revoke_keys: Default::default(),
            },
            balance_overrides: Default::default(),
            state_overrides: Default::default(),
            key: Some(key.to_call_key()),
        })
        .await?;

    let mut value = serde_json::to_value(response)?;
    sort_json_keys(&mut value);
    // TODO: fee related redactions can be removed if we make the tip block of Anvil at the moment
    // of this assertion consistent across runs, right now it's either 15 or 16, and this changes
    // the EIP-1559 fee estimations.
    insta::assert_json_snapshot!(value, {
        ".capabilities.assetDiffs.*" => insta::sorted_redaction(),
        ".capabilities.feeTotals.*.value" => reduction_from_str::<f64>("value"),
        ".context.quote.hash" => reduction_from_str::<B256>("hash"),
        ".context.quote.quotes[].intent.encodedPreCalls[]" => reduction_from_str::<Bytes>("encodedPreCall"),
        ".context.quote.quotes[].intent.prePaymentAmount" => reduction_from_str::<U256>("prePaymentAmount"),
        ".context.quote.quotes[].intent.prePaymentMaxAmount" => reduction_from_str::<U256>("prePaymentMaxAmount"),
        ".context.quote.quotes[].intent.totalPaymentAmount" => reduction_from_str::<U256>("totalPaymentAmount"),
        ".context.quote.quotes[].intent.totalPaymentMaxAmount" => reduction_from_str::<U256>("totalPaymentMaxAmount"),
        ".context.quote.quotes[].nativeFeeEstimate.maxFeePerGas" => reduction_from_str::<U256>("maxFeePerGas"),
        ".context.quote.r" => reduction_from_str::<U256>("r"),
        ".context.quote.s" => reduction_from_str::<U256>("s"),
        ".context.quote.ttl" => insta::dynamic_redaction(|value, _path| {
            assert!(value.as_u64().unwrap() > std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs());
            "[ttl]"
        }),
        ".context.quote.v" => reduction_from_str::<U64>("v"),
        ".context.quote.yParity" => reduction_from_str::<U64>("yParity"),
        ".digest" => reduction_from_str::<B256>("digest"),
        ".typedData.message.encodedPreCalls[]" => reduction_from_str::<Bytes>("encodedPreCall"),
        ".typedData.message.prePaymentMaxAmount" => reduction_from_str::<U256>("prePaymentMaxAmount"),
        ".typedData.message.totalPaymentMaxAmount" => reduction_from_str::<U256>("totalPaymentMaxAmount"),
    });

    Ok(())
}

#[tokio::test]
async fn test_prepare_upgrade_account() -> eyre::Result<()> {
    let env = Environment::setup().await?;

    let admin_key = KeyWith712Signer::mock_admin_with_key(KeyType::Secp256k1, ADMIN_KEY)?.unwrap();

    let response = env
        .relay_endpoint
        .prepare_upgrade_account(PrepareUpgradeAccountParameters {
            address: env.eoa.address(),
            delegation: env.delegation,
            chain_id: None,
            capabilities: UpgradeAccountCapabilities {
                authorize_keys: vec![admin_key.to_authorized()],
            },
        })
        .await?;

    // Nonces are random, so we need to redact them and the digest that depends on them.
    insta::assert_json_snapshot!(response, {
        ".digests.exec" => reduction_from_str::<B256>("digest"),
        ".context.preCall.nonce" => reduction_from_str::<B256>("nonce"),
        ".typedData.message.nonce" => reduction_from_str::<B256>("nonce"),
    });

    Ok(())
}

#[tokio::test]
async fn test_send_prepared_calls() -> eyre::Result<()> {
    let config =
        EnvironmentConfig { num_chains: 2, fee_recipient: Address::ZERO, ..Default::default() };
    let env = Environment::setup_with_config(config.clone()).await?;

    // Create a key for signing
    let key = KeyWith712Signer::mock_admin_with_key(KeyType::Secp256k1, ADMIN_KEY)?.unwrap();

    // Account upgrade deployed onchain.
    upgrade_account_lazily(&env, &[key.to_authorized()], AuthKind::Auth).await?;

    let balance =
        IERC20::new(env.erc20, env.provider_for(1)).balanceOf(env.eoa.address()).call().await?
            / uint!(2_U256);
    let response = env
        .relay_endpoint
        .prepare_calls(PrepareCallsParameters {
            calls: vec![Call::transfer(env.erc20, Address::ZERO, uint!(1_U256))],
            chain_id: env.chain_id_for(0),
            from: Some(env.eoa.address()),
            capabilities: PrepareCallsCapabilities {
                authorize_keys: Default::default(),
                meta: Meta { fee_token: env.erc20, fee_payer: None, nonce: Some(U256::ZERO) },
                pre_calls: Default::default(),
                pre_call: Default::default(),
                required_funds: vec![RequiredAsset::new(env.erc20, balance)],
                revoke_keys: Default::default(),
            },
            balance_overrides: Default::default(),
            state_overrides: Default::default(),
            key: Some(key.to_call_key()),
        })
        .await?;

    let signature = key.sign_payload_hash(response.digest).await?;

    let response = env
        .relay_endpoint
        .send_prepared_calls(SendPreparedCallsParameters {
            capabilities: Default::default(),
            context: response.context,
            key: key.to_call_key(),
            signature,
        })
        .await?;

    insta::assert_json_snapshot!(response, {
        ".id" => reduction_from_str::<BundleId>("id"),
    });

    Ok(())
}

#[tokio::test]
async fn test_upgrade_account() -> eyre::Result<()> {
    let env = Environment::setup().await?;

    let admin_key = KeyWith712Signer::mock_admin_with_key(KeyType::Secp256k1, ADMIN_KEY)?.unwrap();

    let response = env
        .relay_endpoint
        .prepare_upgrade_account(PrepareUpgradeAccountParameters {
            address: env.eoa.address(),
            delegation: env.delegation,
            chain_id: None,
            capabilities: UpgradeAccountCapabilities {
                authorize_keys: vec![admin_key.to_authorized()],
            },
        })
        .await?;

    // Sign Intent digest
    let precall_signature = env.eoa.sign_hash(&response.digests.exec).await?;

    // Sign 7702 delegation
    let nonce = env.provider().get_transaction_count(env.eoa.address()).await?;
    let authorization = AuthKind::Auth.sign(&env, nonce).await?;

    // Upgrade account.
    #[allow(clippy::let_unit_value)]
    let response = env
        .relay_endpoint
        .upgrade_account(UpgradeAccountParameters {
            context: response.context,
            signatures: UpgradeAccountSignatures {
                auth: authorization.signature()?,
                exec: precall_signature,
            },
        })
        .await?;

    insta::assert_json_snapshot!(response);

    Ok(())
}

#[tokio::test]
async fn test_get_authorization() -> eyre::Result<()> {
    let env = Environment::setup().await?;

    let admin_key = KeyWith712Signer::mock_admin_with_key(KeyType::Secp256k1, ADMIN_KEY)?.unwrap();

    let response = env
        .relay_endpoint
        .prepare_upgrade_account(PrepareUpgradeAccountParameters {
            address: env.eoa.address(),
            delegation: env.delegation,
            chain_id: None,
            capabilities: UpgradeAccountCapabilities {
                authorize_keys: vec![admin_key.to_authorized()],
            },
        })
        .await?;

    // Sign Intent digest
    let precall_signature = env.eoa.sign_hash(&response.digests.exec).await?;

    // Sign 7702 delegation
    let nonce = env.provider().get_transaction_count(env.eoa.address()).await?;
    let authorization = AuthKind::Auth.sign(&env, nonce).await?;

    // Upgrade account.
    env.relay_endpoint
        .upgrade_account(UpgradeAccountParameters {
            context: response.context,
            signatures: UpgradeAccountSignatures {
                auth: authorization.signature()?,
                exec: precall_signature,
            },
        })
        .await?;

    // Get authorization
    let response = env
        .relay_endpoint
        .get_authorization(GetAuthorizationParameters { address: env.eoa.address() })
        .await?;

    insta::assert_json_snapshot!(response);

    Ok(())
}

#[tokio::test]
async fn test_get_calls_status() -> eyre::Result<()> {
    let config =
        EnvironmentConfig { num_chains: 2, fee_recipient: Address::ZERO, ..Default::default() };
    let env = Environment::setup_with_config(config.clone()).await?;

    // Create a key for signing
    let key = KeyWith712Signer::mock_admin_with_key(KeyType::Secp256k1, ADMIN_KEY)?.unwrap();

    // Account upgrade deployed onchain.
    upgrade_account_lazily(&env, &[key.to_authorized()], AuthKind::Auth).await?;

    let balance =
        IERC20::new(env.erc20, env.provider_for(1)).balanceOf(env.eoa.address()).call().await?
            / uint!(2_U256);
    let response = env
        .relay_endpoint
        .prepare_calls(PrepareCallsParameters {
            calls: vec![Call::transfer(env.erc20, Address::ZERO, uint!(1_U256))],
            chain_id: env.chain_id_for(0),
            from: Some(env.eoa.address()),
            capabilities: PrepareCallsCapabilities {
                authorize_keys: Default::default(),
                meta: Meta { fee_token: env.erc20, fee_payer: None, nonce: Some(U256::ZERO) },
                pre_calls: Default::default(),
                pre_call: Default::default(),
                required_funds: vec![RequiredAsset::new(env.erc20, balance)],
                revoke_keys: Default::default(),
            },
            balance_overrides: Default::default(),
            state_overrides: Default::default(),
            key: Some(key.to_call_key()),
        })
        .await?;

    let signature = key.sign_payload_hash(response.digest).await?;

    let response = env
        .relay_endpoint
        .send_prepared_calls(SendPreparedCallsParameters {
            capabilities: Default::default(),
            context: response.context,
            key: key.to_call_key(),
            signature,
        })
        .await?;

    await_calls_status(&env, response.id).await?;

    let response = env.relay_endpoint.get_calls_status(response.id).await?;

    // TODO: these redactions are due to the same issue as in `test_prepare_calls`
    insta::assert_json_snapshot!(response, {
        ".id" => insta::dynamic_redaction(move |value, _path| {
            assert_eq!(BundleId::from_str(value.as_str().unwrap()), Ok(response.id));
            "[id]"
        }),
        ".receipts[].logs[].topics[]" => reduction_from_str::<B256>("topic"),
        ".receipts[].logs[].data" => reduction_from_str::<Bytes>("data"),
        ".receipts[].logs[].blockHash" => reduction_from_str::<B256>("blockHash"),
        ".receipts[].logs[].blockNumber" => reduction_from_str::<U64>("blockNumber"),
        ".receipts[].logs[].blockTimestamp" => reduction_from_str::<U64>("blockTimestamp"),
        ".receipts[].logs[].transactionHash" => reduction_from_str::<B256>("transactionHash"),
        ".receipts[].blockHash" => reduction_from_str::<B256>("blockHash"),
        ".receipts[].blockNumber" => reduction_from_str::<U64>("blockNumber"),
        ".receipts[].gasUsed" => reduction_from_str::<U64>("gasUsed"),
        ".receipts[].transactionHash" => reduction_from_str::<B256>("transactionHash"),
    });

    Ok(())
}

#[tokio::test]
async fn test_verify_signature() -> eyre::Result<()> {
    let env = Environment::setup().await?;

    let key = KeyWith712Signer::mock_admin_with_key(KeyType::Secp256k1, ADMIN_KEY)?.unwrap();
    upgrade_account_lazily(&env, &[key.to_authorized()], AuthKind::Auth).await?;

    let digest = B256::ZERO;
    let signature = key.sign_payload_hash(digest).await?;

    let response = env
        .relay_endpoint
        .verify_signature(VerifySignatureParameters {
            address: env.eoa.address(),
            chain_id: env.chain_id(),
            digest,
            signature: signature.clone(),
        })
        .await?;

    insta::assert_json_snapshot!(response, {
        ".proof.initPreCall.nonce" => reduction_from_str::<U256>("nonce"),
        ".proof.initPreCall.signature" => reduction_from_str::<Bytes>("signature"),
    });

    Ok(())
}

#[tokio::test]
async fn test_add_faucet_funds() -> eyre::Result<()> {
    let env = Environment::setup().await?;

    let response = env
        .relay_endpoint
        .add_faucet_funds(AddFaucetFundsParameters {
            token_address: env.fee_token,
            address: env.eoa.address(),
            chain_id: env.chain_id(),
            value: U256::ONE,
        })
        .await?;

    insta::assert_json_snapshot!(response, {
        ".transactionHash" => reduction_from_str::<B256>("transactionHash")
    });

    Ok(())
}

/// Creates a reduction that asserts value can be parsed using [`FromStr`], and replaces
/// the value with the provided name.
fn reduction_from_str<T: FromStr>(name: &str) -> insta::internals::Redaction {
    let name = name.to_string();
    insta::dynamic_redaction(move |value, _path| {
        assert!(T::from_str(value.as_str().unwrap()).is_ok());
        format!("[{}]", name)
    })
}

fn sort_json_keys(v: &mut serde_json::Value) {
    use serde_json::Value;
    match v {
        Value::Object(map) => {
            map.sort_keys();
        }
        Value::Array(arr) => {
            for v in arr.iter_mut() {
                sort_json_keys(v);
            }
        }
        _ => {}
    }
}
