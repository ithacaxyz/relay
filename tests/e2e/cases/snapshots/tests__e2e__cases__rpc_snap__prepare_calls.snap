---
source: tests/e2e/cases/rpc_snap.rs
expression: value
---
{
  "capabilities": {
    "assetDiffs": {
      "0x7a69": [
        [
          "0x0000000000000000000000000000000000000000",
          [
            {
              "address": "0xed1db453c3156ff3155a97ad217b3087d5dc5f6e",
              "decimals": 18,
              "direction": "incoming",
              "fiat": {
                "currency": "usd",
                "value": "0.000000000000000002"
              },
              "name": "mockName",
              "recipients": [],
              "symbol": "mockSymbol",
              "type": "erc20",
              "value": "0x2"
            }
          ]
        ],
        [
          "0xa023ba34fe23ad51d00515a68657e232ab6a1297",
          [
            {
              "address": "0xed1db453c3156ff3155a97ad217b3087d5dc5f6e",
              "decimals": 18,
              "direction": "outgoing",
              "fiat": {
                "currency": "usd",
                "value": "0.000000000000000001"
              },
              "name": "mockName",
              "recipients": [
                "0x0000000000000000000000000000000000000000"
              ],
              "symbol": "mockSymbol",
              "type": "erc20",
              "value": "0x1"
            }
          ]
        ]
      ]
    },
    "authorizeKeys": [],
    "feeTotals": {
      "0x0": {
        "currency": "usd",
        "value": "[value]"
      },
      "0x7a69": {
        "currency": "usd",
        "value": "[value]"
      }
    },
    "revokeKeys": []
  },
  "context": {
    "quote": {
      "hash": "[hash]",
      "multiChainRoot": null,
      "quotes": [
        {
          "authorizationAddress": "0x29a79095352a718b3d7fe84e1f14e9f34a35598e",
          "chainId": "0x7a69",
          "ethPrice": "0xde0b6b3a7640000",
          "extraPayment": "0x0",
          "feeTokenDeficit": "0x0",
          "intent": {
            "combinedGas": "[combinedGas]",
            "encodedFundTransfers": [],
            "encodedPreCalls": [
              "[encodedPreCall]"
            ],
            "eoa": "0xa023ba34fe23ad51d00515a68657e232ab6a1297",
            "executionData": "0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000020000000000000000000000000ed1db453c3156ff3155a97ad217b3087d5dc5f6e000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000044a9059cbb0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000",
            "expiry": "0x0",
            "funder": "0x0000000000000000000000000000000000000000",
            "funderSignature": "0x",
            "isMultichain": false,
            "nonce": "0x0",
            "payer": "0x0000000000000000000000000000000000000000",
            "paymentAmount": "[paymentMaxAmount]",
            "paymentMaxAmount": "[paymentMaxAmount]",
            "paymentRecipient": "0x0000000000000000000000000000000000000000",
            "paymentSignature": "0x",
            "paymentToken": "0xed1db453c3156ff3155a97ad217b3087d5dc5f6e",
            "settler": "0x0000000000000000000000000000000000000000",
            "settlerContext": "0x",
            "signature": "0x",
            "supportedAccountImplementation": "0xb19b36b1456e65e3a6d514d3f715f204bd59f431"
          },
          "nativeFeeEstimate": {
            "maxFeePerGas": "[maxFeePerGas]",
            "maxPriorityFeePerGas": "0x1"
          },
          "orchestrator": "0x700b6a60ce7eaaea56f065753d8dcb9653dbad35",
          "paymentTokenDecimals": 18,
          "txGas": "[txGas]"
        }
      ],
      "r": "[r]",
      "s": "[s]",
      "ttl": "[ttl]",
      "v": "[v]",
      "yParity": "[yParity]"
    }
  },
  "digest": "[digest]",
  "key": {
    "prehash": false,
    "publicKey": "0x0000000000000000000000001409e2aad496505095defaaa2baff67c65168e65",
    "type": "secp256k1"
  },
  "typedData": {
    "domain": {
      "chainId": "0x7a69",
      "name": "Orchestrator",
      "verifyingContract": "0x700b6a60ce7eaaea56f065753d8dcb9653dbad35",
      "version": "0.5.3"
    },
    "message": {
      "calls": [
        {
          "data": "0xa9059cbb00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001",
          "to": "0xed1db453c3156ff3155a97ad217b3087d5dc5f6e",
          "value": "0x0"
        }
      ],
      "combinedGas": "[combinedGas]",
      "encodedFundTransfers": [],
      "encodedPreCalls": [
        "[encodedPreCall]"
      ],
      "eoa": "0xa023ba34fe23ad51d00515a68657e232ab6a1297",
      "expiry": "0x0",
      "multichain": false,
      "nonce": "0x0",
      "payer": "0x0000000000000000000000000000000000000000",
      "paymentMaxAmount": "[paymentMaxAmount]",
      "paymentToken": "0xed1db453c3156ff3155a97ad217b3087d5dc5f6e",
      "settler": "0x0000000000000000000000000000000000000000"
    },
    "primaryType": "Intent",
    "types": {
      "Call": [
        {
          "name": "to",
          "type": "address"
        },
        {
          "name": "value",
          "type": "uint256"
        },
        {
          "name": "data",
          "type": "bytes"
        }
      ],
      "EIP712Domain": [
        {
          "name": "name",
          "type": "string"
        },
        {
          "name": "version",
          "type": "string"
        },
        {
          "name": "chainId",
          "type": "uint256"
        },
        {
          "name": "verifyingContract",
          "type": "address"
        }
      ],
      "Intent": [
        {
          "name": "multichain",
          "type": "bool"
        },
        {
          "name": "eoa",
          "type": "address"
        },
        {
          "name": "calls",
          "type": "Call[]"
        },
        {
          "name": "nonce",
          "type": "uint256"
        },
        {
          "name": "payer",
          "type": "address"
        },
        {
          "name": "paymentToken",
          "type": "address"
        },
        {
          "name": "paymentMaxAmount",
          "type": "uint256"
        },
        {
          "name": "combinedGas",
          "type": "uint256"
        },
        {
          "name": "encodedPreCalls",
          "type": "bytes[]"
        },
        {
          "name": "encodedFundTransfers",
          "type": "bytes[]"
        },
        {
          "name": "settler",
          "type": "address"
        },
        {
          "name": "expiry",
          "type": "uint256"
        }
      ]
    }
  }
}
